lib/appsignal/tracer.ex

```diff
@@ -258,4 +261,38 @@ defmodule Appsignal.Tracer do
       ArgumentError -> nil
     end
   end
+
+  @spec on_create_span(Span.t() | nil, Span.t() | nil) :: Span.t() | nil
+  defp on_create_span(span, parent) do
+    custom_on_create_fun =
+      Application.get_env(:appsignal, :custom_on_create_fun, &__MODULE__.custom_on_create_fun/2)
+
+    span = custom_on_create_fun.(span, parent)
+    span
+  end
+
+  @doc """
+  This function can be defined by the user and will be executed on the
+  creation of the span after create_span/3 is executed. It can be used to add
+  custom_data to the span.
+
+  Example in your own application:
+  ```ex
+  defmodule MyApp.Appsignal do
+    def custom_on_create_fun(span, parent) do
+      Appsignal.Span.set_sample_data(span, "custom_data", %{"foo": "bar"})
+    end
+  end
+  ```
+
+  This can be added to the config with:
+  ```ex
+  config :appsignal, custom_on_create_fun: &MyApp.Appsignal.custom_on_create_fun/2
+  ```
+  """
+
+  @spec custom_on_create_fun(Span.t() | nil, Span.t() | nil) :: Span.t() | nil
+  def custom_on_create_fun(span, _parent) do
+    span
+  end
```

Comment:
I think the return value of the custom function should be ignored, not used as the new span to be returned by `create_span`.

This is what we do with `instrument/2`, for example: you pass a function to it that takes the span as an argument, and that function can use that span to call `Appsignal.Span` functions, but it doesn't expect the span to be returned.

```suggestion
  @spec custom_on_create_fun(Span.t() | nil, Span.t() | nil) :: any()
  def custom_on_create_fun(_span, _parent) do
    nil
  end
```

(aside from the suggestion diff here, `on_create_span`, it should not reassign `span` to the return value of the `custom_on_create_fun` that it calls)